lib/reversetunnel/agent.go

```diff
@@ -22,502 +22,541 @@ package reversetunnel
 	}
-	if a.Client == nil {
-		return trace.BadParameter("missing parameter Client")
+	if c.sshDialer == nil {
+		return trace.BadParameter("missing parameter sshDialer")
 	}
-	if a.AccessPoint == nil {
-		return trace.BadParameter("missing parameter AccessPoint")
+	if c.transporter == nil {
+		return trace.BadParameter("missing parameter transporter")
 	}
-	if a.Signer == nil {
-		return trace.BadParameter("missing parameter Signer")
+	if c.tracker == nil {
+		return trace.BadParameter("missing parameter tracker")
 	}
-	if len(a.Username) == 0 {
-		return trace.BadParameter("missing parameter Username")
+	if c.clock == nil {
+		c.clock = clockwork.NewRealClock()
 	}
-	if a.Clock == nil {
-		a.Clock = clockwork.NewRealClock()
+	if c.log == nil {
+		c.log = logrus.New()
 	}
-	logger := a.Log
-	if a.Log == nil {
-		logger = log.StandardLogger()
+	if !c.lease.IsZero() {
+		c.log = c.log.WithField("leaseID", c.lease.ID())
 	}
-	a.Log = logger.WithFields(log.Fields{
-		trace.Component: teleport.Component(a.Component, teleport.ComponentReverseTunnelAgent),
-		trace.ComponentFields: log.Fields{
-			"target":  a.Addr.String(),
-			"leaseID": a.Lease.ID(),
-		},
-	})
-	return nil
-}
 
-// Agent is a reverse tunnel agent running as a part of teleport Proxies
-// to establish outbound reverse tunnels to remote proxies.
-//
-// There are two operation modes for agents:
-// * Standard agent attempts to establish connection
-// to any available proxy. Standard agent transitions between
-// "connecting" -> "connected states.
-// * Discovering agent attempts to establish connection to a subset
-// of remote proxies (specified in the config via DiscoverProxies parameter.)
-// Discovering agent transitions between "discovering" -> "discovered" states.
-type Agent struct {
-	sync.RWMutex
-	AgentConfig
-	log         log.FieldLogger
-	ctx         context.Context
-	cancel      context.CancelFunc
-	authMethods []ssh.AuthMethod
-	// state is the state of this agent
-	state string
-	// stateChange records last time the state was changed
-	stateChange time.Time
-	// principals is the list of principals of the server this agent
-	// is currently connected to
-	principals []string
+	c.log = c.log.WithField("target", c.addr.String())
+
+	return nil
 }
 
-// ReverseTunnelDetails contains catchable details about the reverse tunnel.
-type reverseTunnelDetails struct {
-	// TLSRoutingEnabled indicates that remote address listener supports ALPN SNI Listener and
-	// the client needs to dial the remote proxy with proper TLS ALPN protocol.
-	TLSRoutingEnabled bool
+// agent creates and manages a reverse tunnel to a remote proxy server.
+type agent struct {
+	agentConfig
+	// client is a client for the agent's ssh connection.
+	client SSHClient
+	// state is the internal state of an agent. Use GetState for threadsafe access.
+	state AgentState
+	// doneConnecting is closed when an agent's state will never transition to connecting.
+	doneConnecting chan struct{}
```

Comment:
Nit: I think its important to spell out the fact that `doneConnecting` is a memory barrier that prevents unsynchronized access to various fields and protects the `WaitGroup`s from having `.Wait()` called too early.  Feels a bit too easy to overlook the importance of this field when refactoring and misuse it.